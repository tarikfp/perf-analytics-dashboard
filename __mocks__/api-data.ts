export const API_DATA = [
  {
    _id: '62704c195a3083ec7d7a3c08',
    ttfb: 0.011,
    fcp: 0.37559999999403954,
    domLoad: 0.293,
    windowLoad: 0.406,
    resources: [
      {
        duration: 7,
        initiatorType: 'script',
        name: 'http://localhost:3000/@vite/client',
        responseEnd: 26.400000005960464,
        transferSize: 300,
        startTime: 19.400000005960464,
      },
      {
        duration: 37.5,
        initiatorType: 'script',
        name: 'http://localhost:3000/src/main.tsx?t=1651526659795',
        responseEnd: 57.099999994039536,
        transferSize: 2517,
        startTime: 19.599999994039536,
      },
    ],
    userAgent:
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.127 Safari/537.36',
    url: 'http://localhost:3000/',
    createdAt: '2022-05-02T21:24:41.396Z',
    __v: 0,
  },
];
