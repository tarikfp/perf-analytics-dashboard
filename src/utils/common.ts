/* eslint-disable arrow-body-style */
import { PerformanceResourceTimingHandler } from '@/types';

// helpers

const convertToSec = (value: number): number => value / 1000;

const takeFirstDigits = (value: number): number => Number(value.toFixed(3));

const findDisplayNameFromKey = <T>(
  data: T,
  arr: Array<{ id: keyof T | string; display: string }>,
): string[] => {
  const keys: string[] = [];
  arr.forEach((x) => {
    if (Object.keys(data).includes(x.id as string)) {
      keys.push(x.display);
    }
  });
  return keys;
};

const parsePerformanceResourceTimingData = (
  data: PerformanceResourceTimingHandler[],
): PerformanceResourceTimingHandler[] => {
  return data.map((item) => ({
    ...item,
    responseEnd: takeFirstDigits(convertToSec(item.responseEnd)),
    duration: takeFirstDigits(convertToSec(item.duration)),
    startTime: takeFirstDigits(convertToSec(item.startTime)),
  }));
};

const sortByDate = <T>(data: T[], sortKey: keyof T) => {
  return data.sort((a, b) => {
    const dateA = new Date(Number(a[sortKey]));
    const dateB = new Date(Number(b[sortKey]));
    return dateB.getTime() - dateA.getTime();
  });
};

export {
  sortByDate,
  takeFirstDigits,
  findDisplayNameFromKey,
  parsePerformanceResourceTimingData,
};
