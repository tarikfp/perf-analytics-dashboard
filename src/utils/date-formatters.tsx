import * as dayjs from 'dayjs';

// date formatters

const formatTooltipDate = (date: string): string =>
  dayjs(date).format('YYYY/MM/DD HH:mm:ss');

const formatAxisDate = (date: string): string => dayjs(date).format('HH:mm:ss');

const formatParamDate = (date: Date): string =>
  dayjs(date).format('YYYY-MM-DD HH:mm');

const formatTableDate = (date: Date | number): string =>
  dayjs(date).format('YYYY-MM-DD HH:mm:ss');

export { formatAxisDate, formatParamDate, formatTableDate, formatTooltipDate };
