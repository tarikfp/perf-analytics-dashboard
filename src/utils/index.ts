import dayjs from 'dayjs';

const formatTooltipDate = (date: string): string =>
  dayjs(date).format('YYYY/MM/DD HH:mm:ss');

const formatAxisDate = (date: string): string => dayjs(date).format('HH:mm:ss');

export { formatAxisDate, formatTooltipDate };
