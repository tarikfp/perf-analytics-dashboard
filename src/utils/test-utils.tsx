import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ThemeContext } from '../theme';

const themeContextValue = {
  isDarkMode: false,
  toggleDarkMode: () => {},
};

// helpers to create test component with theme context

export function createWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): renderer.ReactTestRendererJSON | renderer.ReactTestRendererJSON[] | null {
  return renderer
    .create(
      <ThemeContext.Provider value={themeContextValue}>
        {children}
      </ThemeContext.Provider>,
    )
    .toJSON();
}

export function wrapWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): React.ReactElement {
  return (
    <ThemeContext.Provider value={themeContextValue}>
      {children}
    </ThemeContext.Provider>
  );
}
