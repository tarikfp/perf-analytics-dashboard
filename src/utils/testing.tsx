import { render, RenderResult } from '@testing-library/react';
import { ThemeContext } from '@theme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

const themeContextValue = {
  isDarkMode: false,
  toggleDarkMode: () => null,
};

// helpers to create test component with theme context

export function createWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): renderer.ReactTestRendererJSON | renderer.ReactTestRendererJSON[] | null {
  return renderer
    .create(
      <ThemeContext.Provider value={themeContextValue}>
        {children}
      </ThemeContext.Provider>,
    )
    .toJSON();
}

export function renderWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): RenderResult {
  return render(
    <ThemeContext.Provider value={themeContextValue}>
      {children}
    </ThemeContext.Provider>,
  );
}

export function wrapWithThemeContext(
  children: React.ReactElement | React.ReactNode,
): React.ReactElement {
  return (
    <ThemeContext.Provider value={themeContextValue}>
      {children}
    </ThemeContext.Provider>
  );
}
