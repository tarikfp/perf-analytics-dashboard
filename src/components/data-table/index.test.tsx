import * as React from 'react';
import DataTable from '.';
import { API_DATA } from '../../../__mocks__/api-data';
import { createWithThemeContext } from '../../utils/test-utils';

const props = {
  data: API_DATA,
};

test('data table renders correctly', () => {
  const tree = createWithThemeContext(<DataTable {...props} />);
  expect(tree).toMatchSnapshot();
});
