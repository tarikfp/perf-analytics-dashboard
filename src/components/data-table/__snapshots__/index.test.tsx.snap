// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data table renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-13p2epw-MuiPaper-root"
>
  <div
    className="MuiTableContainer-root css-102rwc2-MuiTableContainer-root"
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root css-1s6ny8r-MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "backgroundColor": "#ffffff",
              }
            }
          />
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "backgroundColor": "#ffffff",
                "color": "#29434e",
                "minWidth": 100,
              }
            }
          >
            URL
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "backgroundColor": "#ffffff",
                "color": "#29434e",
                "minWidth": 100,
              }
            }
          >
            TTFB
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "backgroundColor": "#ffffff",
                "color": "#29434e",
                "minWidth": 100,
              }
            }
          >
            FCP
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "backgroundColor": "#ffffff",
                "color": "#29434e",
                "minWidth": 100,
              }
            }
          >
            Dom load
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "backgroundColor": "#ffffff",
                "color": "#29434e",
                "minWidth": 100,
              }
            }
          >
            Window load
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignCenter MuiTableCell-sizeMedium css-dwuj3p-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "backgroundColor": "#ffffff",
                "color": "#29434e",
                "minWidth": undefined,
              }
            }
          >
            User agent
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
            scope="col"
            style={
              Object {
                "backgroundColor": "#ffffff",
                "color": "#29434e",
                "minWidth": 100,
              }
            }
          >
            Created At
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-hover css-1q1u3t4-MuiTableRow-root"
          role="checkbox"
          tabIndex={-1}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <button
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                />
              </svg>
            </button>
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            http://localhost:3000/
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            0.011
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            0.376
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            0.293
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            0.406
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
            style={
              Object {
                "minWidth": undefined,
              }
            }
          >
            Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.127 Safari/537.36
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            2022-05-03 00:24:41
          </td>
        </tr>
        <tr
          className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            colSpan={6}
            style={
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              }
            }
          />
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
    >
      <div
        className="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
      />
      <p
        className="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
        id=":r1:"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
        onClick={[Function]}
        variant="standard"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r1: :r0:"
          className="MuiTablePagination-select MuiSelect-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
          id=":r0:"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          10
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value={10}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTablePagination-selectIcon MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
      >
        1–1 of 1
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Go to previous page"
          className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Go to previous page"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
            />
          </svg>
        </button>
        <button
          aria-label="Go to next page"
          className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Go to next page"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
